### ISSUE TITLE
[REVIEW]: Bézier

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#267]
**Submitting author:** @dhermes (<a href="http://orcid.org/0000-0001-7366-173X">Daniel Hermes</a>)
**Repository:** <a href="https://github.com/dhermes/bezier" target ="_blank">https://github.com/dhermes/bezier</a>
**Version:** 0.3.0
**Editor:** @arfon
**Reviewer:** @tacaswell
**Archive:** <a href="http://dx.doi.org/10.5281/zenodo.838308" target="_blank">10.5281/zenodo.838308</a>

## Status

[![status](http://joss.theoj.org/papers/e40b3701ae4f96e1da1caedddafa121a/status.svg)](http://joss.theoj.org/papers/e40b3701ae4f96e1da1caedddafa121a)

Status badge code:

```
HTML: <a href="http://joss.theoj.org/papers/e40b3701ae4f96e1da1caedddafa121a"><img src="http://joss.theoj.org/papers/e40b3701ae4f96e1da1caedddafa121a/status.svg"></a>
Markdown: [![status](http://joss.theoj.org/papers/e40b3701ae4f96e1da1caedddafa121a/status.svg)](http://joss.theoj.org/papers/e40b3701ae4f96e1da1caedddafa121a)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/dhermes/bezier" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

### Conflict of interest

- [x] As the reviewer I confirm that there are no conflicts of interest for me to review this work (such as being a major contributor to the software).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/dhermes/bezier">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [ ] **Version:** Does the release version given match the GitHub release (0.3.0)?
- [x] **Authorship:** Has the submitting author (@dhermes) made major contributions to the software?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [ ] **Performance:** Have any performance claims of the software been confirmed?

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g. API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **References:** Do all archival references that should have a DOI list one (e.g. papers, datasets, software)?

**Comment by whedon** [comment]
Hello human, I'm @whedon. I'm here to help you with some common editorial tasks for JOSS. @tacaswell it looks like you're currently assigned as the reviewer for this paper :tada:.

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As as reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all JOSS reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

**Comment by tacaswell** [comment]
- There are 145 commits to master from the 0.3.0 tag, changes include dependency changes 
 - missing link to docs on the github page (ended up using google to find them)
 - could the `DEVELOPMENT.rst` page be included in the html docs?
 - a bit more narrative as to the applications of this package on the main page would be good

Can not get the functional tests to run following the directions (more-or-less with python36) which I assume is due to changes with absolute import

```
23:03 $ python -m pytest functional_tests/
======================================================================================================== test session starts =========================================================================================================
platform linux -- Python 3.6.1, pytest-3.0.6, py-1.4.32, pluggy-0.4.0
rootdir: /home/tcaswell/src/o/bezier, inifile: tox.ini
plugins: cov-2.4.0, xdist-1.15.0
collected 0 items / 4 errors 

=============================================================================================================== ERRORS ===============================================================================================================
_______________________________________________________________________________________ ERROR collecting functional_tests/test_curve_curve.py ________________________________________________________________________________________
ImportError while importing test module '/home/tcaswell/src/o/bezier/functional_tests/test_curve_curve.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
functional_tests/test_curve_curve.py:20: in <module>
    import runtime_utils
E   ModuleNotFoundError: No module named 'runtime_utils'
_______________________________________________________________________________________ ERROR collecting functional_tests/test_segment_box.py ________________________________________________________________________________________
ImportError while importing test module '/home/tcaswell/src/o/bezier/functional_tests/test_segment_box.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
functional_tests/test_segment_box.py:16: in <module>
    import runtime_utils
E   ModuleNotFoundError: No module named 'runtime_utils'
______________________________________________________________________________________ ERROR collecting functional_tests/test_surface_locate.py ______________________________________________________________________________________
ImportError while importing test module '/home/tcaswell/src/o/bezier/functional_tests/test_surface_locate.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
functional_tests/test_surface_locate.py:17: in <module>
    import runtime_utils
E   ModuleNotFoundError: No module named 'runtime_utils'
_____________________________________________________________________________________ ERROR collecting functional_tests/test_surface_surface.py ______________________________________________________________________________________
ImportError while importing test module '/home/tcaswell/src/o/bezier/functional_tests/test_surface_surface.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
functional_tests/test_surface_surface.py:21: in <module>
    import runtime_utils
E   ModuleNotFoundError: No module named 'runtime_utils'
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 4 errors during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
====================================================================================================== 4 error in 0.83 seconds =======================================================================================================
(dd36) ✘-2 ~/src/o/bezier [0.3.0|✚ 3] 

```

but by changing to functional tests and running ` python -m pytest` got them to pass

Getting one failure on the unit tests
```
(dd36) ✘-1 ~/src/o/bezier [0.3.0|✚ 3] 
23:17 $ python -m pytest tests/
======================================================================================================== test session starts =========================================================================================================
platform linux -- Python 3.6.1, pytest-3.0.6, py-1.4.32, pluggy-0.4.0
rootdir: /home/tcaswell/src/o/bezier, inifile: tox.ini
plugins: cov-2.4.0, xdist-1.15.0
collected 345 items 

tests/test__base.py .......
tests/test__curve_helpers.py ........................
tests/test__helpers.py .......................
tests/test__intersection_helpers.py ..................................................................
tests/test__plot_helpers.py ..
tests/test__surface_helpers.py .........................................................................................................
tests/test_curve.py .......................................
tests/test_curved_polygon.py ..........
tests/test_surface.py .F...................................................................

============================================================================================================== FAILURES ==============================================================================================================
______________________________________________________________________________________________ TestSurface.test___repr__custom_triangle ______________________________________________________________________________________________
Traceback (most recent call last):
  File "/home/tcaswell/src/o/bezier/tests/test_surface.py", line 124, in test___repr__custom_triangle
    nodes = np.zeros((num_nodes, dimension))
TypeError: 'float' object cannot be interpreted as an integer
================================================================================================ 1 failed, 344 passed in 4.10 seconds ================================================================================================
```
Only got part way through this review due to reading about Barycentric coordinate system and learning some new (to me) math!

**Comment by dhermes** [comment]
> There are 145 commits to master from the 0.3.0 tag, changes include dependency changes

Indeed. I have had issues with the Windows build since adding the Fortran speedups. I'd love to do a release but don't want to leave out Windows users.

> missing link to docs on the github page (ended up using google to find them)

There is a badge "docs | latest" in the README. I'm open to suggestions on adding something else / where you'd like to see it added.

![screenshot from 2017-05-23 22-48-11](https://cloud.githubusercontent.com/assets/520669/26388564/eeb9f390-4009-11e7-8f4d-65b365da16be.png)

> could the DEVELOPMENT.rst page be included in the html docs?

Definitely.

> a bit more narrative as to the applications of this package on the main page would be good

- By main page, do you mean http://bezier.readthedocs.io/en/latest/ or the GitHub README?
- When you say "more narrative", do you mean "work with Bezier curves / surfaces / etc." or "here's what a Bezier curve / surface / etc. is and you might want one because <blah>"?

> Can not get the functional tests to run following the directions (more-or-less with python36) which I assume is due to changes with absolute import

Yes I noticed that yesterday when I was updated the CI / testing environment. [Right now][1] they are running in Python 2.7 on CI, but I've filed [an issue][2] to fix and will get to it ASAP tomorrow.

> Getting one failure on the unit tests

I can't reproduce locally or [in CI][3]. Can we discuss your environment a bit more?

> Only got part way through this review due to reading about Barycentric coordinate system and learning some new (to me) math!

They can be really great! One particularly awesome feature of Bezier curves / the Bernstein basis is the convexity property. It allows for lots of nice optimizations. I'm happy to chat about the math if you'd like me to shed some light on things. (The bibliography is a bit long.) I'm certainly not an authority on the subject but there is plenty I do know.

[1]: https://github.com/dhermes/bezier/blob/7a86d904b2e9b48decbd5d6b3fd2340d492b3947/circle.yml#L20
[2]: https://github.com/dhermes/bezier/issues/25
[3]: https://circleci.com/gh/dhermes/bezier

**Comment by tacaswell** [comment]
RE: link to docs:  :sheep: I was looking for a blue link that said 'documentation' and totally skipped over the badges! 

RE: narrative:  both the README and the RTD docs (but content should be very similar).  On the "here's what a Bezier curve / surface / etc. is and you might want one because ..." to address the statement of need.  

----

The test failure is in `def test___repr__custom_triangle(self):` and is raised from the the test code due to

```
(dd36) ✘-1 ~/src/o/bezier [0.3.0|✚ 3] 
08:48 $ ipython
Python 3.6.1 |Continuum Analytics, Inc.| (default, Mar 22 2017, 19:54:23) 
Type "copyright", "credits" or "license" for more information.

IPython 5.2.2 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: import numpy

In [2]: import numpy as np

In [3]: np.zeros((13.0, 7))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-f0a5700cdff1> in <module>()
----> 1 np.zeros((13.0, 7))

TypeError: 'float' object cannot be interpreted as an integer

In [4]: np.__version__
Out[4]: '1.12.1'
```

**Comment by dhermes** [comment]
@tacaswell You're seeing this failure at `HEAD` or at the `0.3.0` tag?

**Comment by dhermes** [comment]
FYI @tacaswell I [added][1] the DEVELOPMENT doc to the HTML / Sphinx docs: http://bezier.readthedocs.io/en/latest/development.html

[1]: https://github.com/dhermes/bezier/commit/8915b040cecb62758b3d7056a1847374d0aa4995

**Comment by dhermes** [comment]
@tacaswell Anything you are waiting on from me?

**Comment by dhermes** [comment]
@tacaswell I have added a [Why Bezier?][1] section and also done a `0.4.0` release.

AFAICT I have addressed all of the current feedback, LMK what I need do from here (e.g. if the `0.4.0` release has thrown a wrench into this process).

[1]: https://bezier.readthedocs.io/en/0.4.0/#why-b-eacute-zier

**Comment by arfon** [comment]
Friendly reminder on this @tacaswell 😁

**Comment by dhermes** [comment]
@tacaswell @arfon Anything I can do here?

**Comment by tacaswell** [comment]
All of my issues have been addressed.

I recommend this to be published!

Sorry for being so slow on my end.

**Comment by tacaswell** [comment]
There are far more references in the bibliography than cited from paper.md, does that get sorted out at somepoint in the next step?

**Comment by arfon** [comment]
Thanks @tacaswell.

> There are far more references in the bibliography than cited from paper.md, does that get sorted out at somepoint in the next step?

@dhermes - all of the references need moving to the `paper.bib` file and citing directly. (You can read how to do that [here](http://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html))

In addition, could you please expand a little on the `paper.md` file? We (the JOSS editors) are in the process of adding more guidance for authors, including a better indication of expected paper length. Currently it's hard to get a picture of the functionality of the software, and also include a statement of need, in less than 250 words. Would you be able to put a little more detail in your submission please?

**Comment by dhermes** [comment]
> could you please expand a little on the `paper.md` file

Will do!

Should I mention features the library **doesn't** have (that could be added at a later date)? Following a standard text / course syllabus I could outline the "toolbox" one would typically want when dealing with Bezier curves and how it compares to the feature set of the library.

----

@arfon IIUC, when @tacaswell said

> There are far more references in the bibliography than cited from `paper.md`

he meant that `paper.bib` has more references (23):

```
$ cat paper.bib | grep -e '^@'
@article{Farouki1987,
@article{Sederberg1989,
@article{Sederberg1989-alg,
@article{Sederberg1990,
@article{Farouki1991,
@article{Dunavant1985,
@article{Farouki1988,
@article{Farouki1990,
@article{Farouki1996,
@article{Kim1998,
@article{Sederberg1986,
@inproceedings{Roth,
@misc{Roth2000,
@article{Farin1986,
@article{Sederberg1986-improp,
@article{Sederberg1984,
@article{Brown1971,
@article{Farouki2012,
@article{Bus2008,
@article{JonssonVavasis,
@techreport{Manocha:CSD-92-698,
@book{Farin2001,
@misc{SederbergNotes,
$ cat paper.bib | grep -e '^@' | wc -l
23
```

than are actually used in `paper.md`:

```
# References

- [Lecture notes][3]: Computer Aided Geometric Design by Thomas W.
  Sederberg of BYU
- [Curves and Surfaces for CAGD][4] by Gerald Farin

...

[3]: http://tom.cs.byu.edu/~557/text/cagd.pdf
[4]: https://www.amazon.com/Curves-Surfaces-CAGD-Fifth-Practical/dp/1558607374
```

**Comment by arfon** [comment]
> Should I mention features the library doesn't have (that could be added at a later date)? Following a standard text / course syllabus I could outline the "toolbox" one would typically want when dealing with Bezier curves and how it compares to the feature set of the library.

I think the paper should only describe the current functionality at a high level rather than write about what the library _might_ do one day.

Also, please note, for the citations, the section after the `# References` heading should be completely empty. Pandoc compiles the references you have cited in your `paper.md` from the `paper.bib` file into the final output. You can use [this paper](https://github.com/arfon/fidgit/blob/master/paper/paper.md) as an example.

**Comment by dhermes** [comment]
I [just updated][1] the summary. LMK what you think @arfon and @tacaswell?

The 3 paragraphs are:

- What are B&#xe9;zier curves / surfaces
- Where (i.e. what fields) have use for them? (This is essentially equivalent to stating the "audience")
- What features are present for people who might use them

[1]: https://github.com/dhermes/bezier/commit/975ac6b1a4313db4dcdc17396d6d34561005939e

**Comment by arfon** [comment]
This looks good to me @dhermes.

**Comment by arfon** [comment]
@dhermes - At this point could you make an archive of the reviewed software in Zenodo/figshare/other service and update this thread with the DOI of the archive? I can then move forward with accepting the submission.

**Comment by dhermes** [comment]
Sure thing: [`10.5281/zenodo.838308`][1] is for `bezier==0.4.0` ([`10.5281/zenodo.838307`][2] is the "catch-all" DOI for the project)

[1]: https://zenodo.org/record/838308
[2]: https://zenodo.org/record/838307

**Comment by arfon** [comment]
@whedon set 10.5281/zenodo.838308 as archive

**Comment by whedon** [comment]
OK. <a href="http://dx.doi.org/10.5281/zenodo.838308" target="_blank">10.5281/zenodo.838308</a> is the archive.

**Comment by arfon** [comment]
@tacaswell many thanks for your review here ✨ 

@dhermes - your paper is now accepted into JOSS and your DOI is http://dx.doi.org/10.21105/joss.00267 ⚡️ 🚀 :boom:

**Comment by dhermes** [comment]
:tada: :boom: w00t! Thanks both of you!
